test/helpers/customError.js

```diff
@@ -1,21 +1,49 @@
-const { config } = require('hardhat');
-
-const optimizationsEnabled = config.solidity.compilers.some(c => c.settings.optimizer.enabled);
+const { expect } = require('chai');
 
 /** Revert handler that supports custom errors. */
-async function expectRevertCustomError(promise, reason) {
+async function expectRevertCustomError(promise, expectedErrorName, args) {
   try {
     await promise;
     expect.fail("Expected promise to throw but it didn't");
   } catch (revert) {
-    if (reason) {
-      if (optimizationsEnabled) {
-        // Optimizations currently mess with Hardhat's decoding of custom errors
-        expect(revert.message).to.include.oneOf([reason, 'unrecognized return data or custom error']);
-      } else {
-        expect(revert.message).to.include(reason);
-      }
+    if (!Array.isArray(args)) {
+      expect.fail('Expected 3rd array parameter for error arguments');
     }
+    // The revert message for custom errors looks like:
+    // VM Exception while processing transaction:
+    // reverted with custom error 'InvalidAccountNonce("0x70997970C51812dc3A010C7d01b50e0d17dc79C8", 0)'
+
+    // We trim out anything inside the single quotes as comma-separated values
+    const [, error] = revert.message.match(/'(.*)'/);
+
+    // Attempt to parse as an error
+    if (!/\w+\(.*\)/g.test(error)) {
+      expect.fail(`Couldn't parse "${error}" as a custom error`);
+    }
+
+    // Extract the error name
+    const [, errorName] = error.match(/(\w+)\(.*\)/);
+
+    const argMatches = [
+      ...error
+        // Remove the error name and leave the parameters
+        .replace(errorName, '')
+        // Capture the remaining parameters
+        .matchAll(/(\w+|-?\d+)/g),
+    ];
```

Comment:
Accepted the suggestion!

> Note that both version will fail if one of the parameters is a string that contains spaces, commas, or other special characters.

This is also correct. I wouldn't worry for that at the moment since I think string is something we're not thinking to return in errors. Could be a guideline though, but we can revisit if we ever need a string with spaces.